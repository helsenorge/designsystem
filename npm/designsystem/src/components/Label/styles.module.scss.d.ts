export type Styles = {
  'after-label-children': string;
  label: string;
  'label--on-dark': string;
  'label--semibold': string;
  'label--sublabel': string;
  'label-content-wrapper': string;
  'label-wrapper': string;
  'label-wrapper--after-label-children': string;
  'label-wrapper--no-bottom-margin': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
