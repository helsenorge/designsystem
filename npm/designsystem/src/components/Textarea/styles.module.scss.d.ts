export type Styles = {
  'input-container': string;
  'input-container__input': string;
  'input-container__input--disabled': string;
  'input-container--bigform': string;
  'input-container--disabled': string;
  'input-container--invalid': string;
  'input-container--on-blueberry': string;
  'input-container--on-dark': string;
  'input-container--transparent': string;
  textarea: string;
  'textarea--gutterBottom': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
