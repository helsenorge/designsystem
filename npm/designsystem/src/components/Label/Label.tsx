import React, { FunctionComponent } from 'react';

import cn from 'classnames';

import { Sublabel, SublabelProps } from './SubLabel';
import { AnalyticsId, FormOnColor } from '../../constants';
import { isComponent } from '../../utils/component';
import FormFieldTag, { FormFieldTagLevel } from '../FormFieldTag';
import Spacer from '../Spacer';
import StatusDot, { StatusDotProps } from '../StatusDot';

import styles from './styles.module.scss';

export type LabelText = {
  hideFromScreenReader?: boolean;
  text: string;
  type?: 'normal' | 'subdued';
};

export type LabelTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'label' | 'p';

export interface LabelProps {
  /** Component shown after label - discourage use of this */
  afterLabelChildren?: React.ReactNode;
  /** Sets the content of the Label */
  children?: React.ReactNode;
  /** Sets a tag that describes whether the form field is required or optional */
  formFieldLevel?: FormFieldTagLevel;
  /** Adds custom classes to the label tag. */
  labelClassName?: string;
  /** Adds custom classes to the label text. */
  labelTextClassName?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Id that is put on the "for" attribute of the label */
  htmlFor?: string;
  /** Changes the underlying element of the label */
  htmlMarkup?: LabelTags;
  /** Id som plasseres p√• <label/> */
  labelId?: string;
  /** Array of main label strings. Can be of type semibold or normal */
  labelTexts?: LabelText[];
  /** Array of sublabel strings. Can be of type semibold or normal */
  onColor?: keyof typeof FormOnColor;
  /** StatusDot placed underneath the last sublabel */
  statusDot?: React.ReactNode;
  /** Sublabel component */
  sublabel?: React.ReactNode;
  /** Adds custom classes to the div wrapping the sublabels. */
  sublabelWrapperClassName?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const getLabelText = (label: React.ReactNode): string => {
  let allLabelText = '';

  if (isComponent<LabelProps>(label, Label)) {
    label.props.labelTexts?.forEach(labelText => {
      allLabelText += !labelText.hideFromScreenReader ? labelText.text : '';
    });
  }

  return allLabelText;
};

export const renderLabel = (label: React.ReactNode, inputId: string, onColor: FormOnColor, markup?: LabelTags): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            htmlMarkup: markup || 'label',
            onColor,
          })
        : typeof label === 'string' && <Label labelTexts={[{ text: label, type: 'normal' }]} htmlFor={inputId} onColor={onColor} />}
    </>
  );
};

export const renderLabelAsParent = (
  label: React.ReactNode,
  children: React.ReactNode,
  inputId: string,
  onColor: FormOnColor,
  labelClassName?: string,
  labelTextClassName?: string,
  sublabelWrapperClassName?: string,
  large?: boolean,
  markup?: LabelTags
): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            onColor,
            children: children,
            labelClassName: cn(labelClassName, label.props.labelClassName),
            labelTextClassName: labelTextClassName,
            htmlMarkup: markup || 'label',
            sublabelWrapperClassName: sublabelWrapperClassName,
            sublabel: large ? undefined : label.props.sublabel,
            statusDot: large ? undefined : label.props.statusDot,
          })
        : typeof label === 'string' && (
            <Label
              labelTexts={[{ text: label, type: 'subdued' }]}
              htmlFor={inputId}
              onColor={onColor}
              htmlMarkup={markup || 'label'}
              labelClassName={labelClassName}
              labelTextClassName={labelTextClassName}
              sublabelWrapperClassName={sublabelWrapperClassName}
            >
              {children}
            </Label>
          )}
    </>
  );
};

const Label: FunctionComponent<LabelProps> = ({
  afterLabelChildren,
  children,
  className,
  formFieldLevel,
  htmlFor,
  htmlMarkup = 'label',
  labelClassName,
  labelTextClassName,
  labelId,
  labelTexts,
  onColor = FormOnColor.onwhite,
  statusDot,
  sublabel,
  sublabelWrapperClassName,
  testId,
}) => {
  const hasChildren = children && typeof children !== 'undefined';
  const labelWrapperClasses = cn(
    styles['label-wrapper'],
    { [styles['label-wrapper--no-bottom-margin']]: hasChildren, [styles['label-wrapper--after-label-children']]: afterLabelChildren },
    className
  );

  const mapLabels = (): React.ReactNode => {
    if (typeof labelTexts === 'undefined') return null;

    return labelTexts.map((labelText, index) => {
      const labelClasses = cn(
        styles.label,
        {
          [styles['label--subdued']]: labelText.type === 'subdued',
          [styles['label--on-dark']]: onColor === FormOnColor.ondark,
        },
        labelTextClassName
      );
      return (
        <span aria-hidden={labelText.hideFromScreenReader} className={labelClasses} key={index}>
          {labelText.text}
        </span>
      );
    });
  };
  const CustomTag = htmlMarkup;

  return (
    <div className={labelWrapperClasses}>
      <div>
        <CustomTag className={labelClassName} id={labelId} htmlFor={htmlFor} data-testid={testId} data-analyticsid={AnalyticsId.Label}>
          <span className={styles['label-content-wrapper']}>
            {children}
            <span className={styles.label__texts}>{mapLabels()}</span>
          </span>
        </CustomTag>
        <div className={sublabelWrapperClassName}>
          {formFieldLevel && <FormFieldTag level={formFieldLevel} />}
          {sublabel &&
            isComponent<SublabelProps>(sublabel, Sublabel) &&
            React.cloneElement(sublabel, {
              onColor: onColor as FormOnColor,
            })}
          {statusDot && isComponent<StatusDotProps>(statusDot, StatusDot) && (
            <>
              <Spacer size={'3xs'} />
              {React.cloneElement(statusDot, {
                onColor: onColor === FormOnColor.ondark ? 'ondark' : 'onwhite',
              })}
            </>
          )}
        </div>
      </div>
      {afterLabelChildren && <div className={styles['after-label-children']}>{afterLabelChildren}</div>}
    </div>
  );
};

export default Label;
