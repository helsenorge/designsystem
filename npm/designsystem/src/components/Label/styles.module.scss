@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';

.label-wrapper {
  margin-bottom: getSpacer(2xs);
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    line-height: $lineheight-size-sm;
  }

  &--no-bottom-margin {
    margin-bottom: 0;
  }

  &--after-label-children {
    display: flex;
    align-items: flex-end;
  }
}

.label-content-wrapper {
  display: flex;
}

.label {
  font-size: $font-size-sm;
  font-weight: 400;
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    font-size: $font-size-md;
    line-height: $lineheight-size-sm;
  }

  &--semibold {
    font-weight: 600;
  }

  &--sublabel {
    display: block;
    font-size: $font-size-xs;
    line-height: 1.25rem;
    color: $neutral800;

    @media (min-width: map.get($grid-breakpoints, sm)) {
      font-size: $font-size-sm;
      line-height: $lineheight-size-xs;
    }
  }

  &--on-dark {
    color: $white;
  }

  &::after {
    content: ' ';
  }
}

.after-label-children {
  padding-left: getSpacer(xs);
}
