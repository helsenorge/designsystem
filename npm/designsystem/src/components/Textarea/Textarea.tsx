import React, { useState, useRef, useEffect } from 'react';
import cn from 'classnames';

import styles from './styles.module.scss';
import { AnalyticsId, AVERAGE_CHARACTER_WIDTH_PX, FormMode } from '../../constants';
import { uuid } from '../../utils/uuid';
import ErrorWrapper from '../ErrorWrapper';

interface TextareaProps
  extends Pick<
    React.InputHTMLAttributes<HTMLTextAreaElement>,
    'autoFocus' | 'disabled' | 'name' | 'autoComplete' | 'placeholder' | 'readOnly' | 'required' | 'defaultValue'
  > {
  /** max character limit in textarea  */
  maxCharacters?: number;
  /** The text is displayed in the end of the text-counter */
  maxText?: string;
  /** Width of textarea in characters (approximate) */
  width?: number;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** If true, the component will have a bottom margin. */
  marginBottom?: boolean;
  /** If true, the component will be transparent. */
  transparent?: boolean;
  /** Changes the visuals of the textarea */
  mode?: keyof typeof FormMode;
  /** Label of the input */
  label?: string;
  /** id of the textarea */
  textareaId?: string;
  /** max rows */
  maxRows?: number;
  /** min rows */
  minRows?: number;
  /** auto-grows until maxRows */
  grow?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Component shown after label */
  afterLabelChildren?: React.ReactNode;
  /** Component shown under label */
  belowLabelChildren?: React.ReactNode;
}

const getTextareaMaxWidth = (characters: number): string => {
  const paddingWidth = '2rem';
  const scrollbarWidth = '16px';
  const borderWidth = '4px';

  return `calc(${characters * AVERAGE_CHARACTER_WIDTH_PX}px + ${paddingWidth} + ${scrollbarWidth} + ${borderWidth})`;
};

const Textarea = React.forwardRef((props: TextareaProps, ref: React.Ref<HTMLTextAreaElement>) => {
  const {
    maxCharacters: max,
    maxText,
    width,
    testId,
    defaultValue,
    marginBottom: gutterBottom,
    transparent,
    mode,
    label,
    textareaId = uuid(),
    minRows = 3,
    maxRows = 10,
    grow,
    errorText,
    afterLabelChildren,
    belowLabelChildren,
    autoFocus,
    disabled,
    name,
    autoComplete,
    placeholder,
    readOnly,
    required,
  } = props;

  const [rows, setRows] = useState(minRows);
  const [textareaInput, setTextareaInput] = useState(defaultValue || '');
  const referanse = useRef<HTMLDivElement>(null);

  const resizeHeight = (target: HTMLTextAreaElement): void => {
    const textareaLineHeight = 28;

    const previousRows = target.rows;
    target.rows = minRows; // reset number of rows in textarea

    const currentRows = Math.floor((target.scrollHeight - 16) / textareaLineHeight); // scrollHeight - 16px of padding to calculate the rows

    if (currentRows === previousRows) {
      target.rows = currentRows;
    }

    if (currentRows >= maxRows) {
      target.rows = maxRows;
      target.scrollTop = target.scrollHeight;
    }

    if (currentRows < maxRows) {
      setRows(currentRows);
    } else {
      setRows(maxRows);
    }
  };

  const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement>): void => {
    if (grow) {
      resizeHeight(event.target);
    }
    setTextareaInput(event.target.value);
  };

  const onDark = mode === FormMode.ondark;
  const onBlueberry = mode === FormMode.onblueberry;
  const textHasError = max && textareaInput.toString().length > max;
  const onError = mode === FormMode.oninvalid || errorText || textHasError;

  const textareaWrapperClass = cn(styles.textarea, {
    [styles['textarea--gutterBottom']]: gutterBottom,
  });

  const labelWrapperClass = cn(styles['textarea__label-wrapper'], {
    [styles[`textarea__label-wrapper--on-dark`]]: onDark,
  });

  const contentWrapperClass = cn(styles['content-wrapper'], {
    [styles['content-wrapper--transparent']]: transparent,
    [styles['content-wrapper--on-blueberry']]: onBlueberry,
    [styles['content-wrapper--on-dark']]: onDark,
    [styles['content-wrapper--invalid']]: onError,
    [styles['content-wrapper--disabled']]: props.disabled,
  });

  const textareaClass = cn(styles['content-wrapper__input'], {
    [styles[`content-wrapper__input--disabled`]]: props.disabled,
  });

  const counterTextClass = cn(styles['textarea__counter-wrapper'], {
    [styles[`textarea__counter-wrapper--on-dark`]]: onDark,
    [styles[`textarea__counter-wrapper--invalid`]]: onError,
  });

  useEffect(() => {
    if (grow && referanse.current?.children && referanse.current?.children[0]) {
      const textarea = referanse.current?.children[0] as HTMLTextAreaElement;
      resizeHeight(textarea);
    }
  }, []);

  let progress = 0;
  if (max) {
    progress = textareaInput.toString().length / max;
  }

  const ariaLevel = progress > 0.95 ? 'polite' : 'off';
  const maxWidth = width ? getTextareaMaxWidth(width) : undefined;

  return (
    <ErrorWrapper errorText={errorText}>
      <div data-testid={testId} data-analyticsid={AnalyticsId.Textarea} className={textareaWrapperClass}>
        {label && (
          <div className={labelWrapperClass}>
            <label htmlFor={textareaId}>{label}</label>
            {afterLabelChildren && <div className={styles['textarea__after-label-children']}>{afterLabelChildren}</div>}
          </div>
        )}
        {belowLabelChildren && <div>{belowLabelChildren}</div>}
        <div className={contentWrapperClass} ref={referanse} style={{ maxWidth }}>
          <textarea
            rows={rows}
            defaultValue={defaultValue}
            id={textareaId}
            className={textareaClass}
            ref={ref}
            onChange={handleChange}
            aria-invalid={!!onError}
            autoFocus={autoFocus}
            disabled={disabled}
            name={name}
            autoComplete={autoComplete}
            placeholder={placeholder}
            readOnly={readOnly}
            required={required}
          />
        </div>
        {max && (
          <div aria-live={ariaLevel} aria-atomic={'true'} className={counterTextClass} style={{ maxWidth }}>
            <p>{`${textareaInput.toString().length}/${max} ${maxText ? maxText : 'tegn'}`}</p>
          </div>
        )}
      </div>
    </ErrorWrapper>
  );
});

export default Textarea;
