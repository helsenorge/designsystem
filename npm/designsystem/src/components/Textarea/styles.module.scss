@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';
@import '../../scss/input';

.textarea {
  @include input-wrapper;

  &--gutterBottom {
    margin-bottom: getSpacer(s);
  }
}

.input-container {
  @include input-container;

  &__input {
    line-height: 28px;
    height: auto;
    padding: 16px;
    resize: none;
  }
}
