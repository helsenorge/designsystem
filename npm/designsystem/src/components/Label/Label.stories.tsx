import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Label, { LabelText } from './Label';
import { Sublabel } from './SubLabel';
import { IconSize } from '../../constants';
import { getColor } from '../../theme/currys';
import Checkbox from '../Checkbox';
import Icon from '../Icon';
import HelpSign from '../Icons/HelpSign';
import Input from '../Input';
import RadioButton from '../RadioButton/RadioButton';
import Select from '../Select';
import StatusDot from '../StatusDot';
import Textarea from '../Textarea';

type LabelWithAndCustomArgs = React.ComponentProps<typeof Label> & {
  showIcon: boolean;
};

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Label',
  component: Label,
  parameters: {
    docs: {
      description: {
        component:
          'Som en innbygger vil jeg kunne se flere dataelementer gruppert i en og samme label på en konsistent måte slik at jeg mer effektivt kan scanne og forstå valgene jeg skal kunne gjøre. \n\n <b>For å sette opp aria-label riktig med Label komponentet:</b> \n\n Gi input feltet sin id til id propen til Label. Gi Input feltet sin aria-describedby prop eventuelle sublabel/statusdot id(er). Du kan skjule bestemte label/sublabel tekster for skjermlesere ved hjelp av hideFromScreenReader property per tekst.',
      },
    },
  },
  argTypes: {},
} satisfies Meta<LabelWithAndCustomArgs>;

export default meta;

type Story = StoryObj<LabelWithAndCustomArgs>;

export const Default: Story = {
  render: args => {
    const sublabelId1 = 'sublabel-testid1';
    const statusDotId1 = 'statusdot-testid1';
    const sublabelId2 = 'sublabel-testid2';
    const statusDotId2 = 'statusdot-testid2';
    const sublabelId3 = 'sublabel-testid3';
    const statusDotId3 = 'statusdot-testid3';

    const sublabelTexts: LabelText[] = [
      { text: 'This is a normal sublabel', type: 'normal' },
      { text: 'Semibold sublabel - hidden from screen reader', type: 'semibold', hideFromScreenReader: true },
    ];

    return (
      <>
        <Input
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
                { text: 'this is hidden from screen readers', hideFromScreenReader: true },
              ]}
              sublabel={<Sublabel id={sublabelId1} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId1} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId1 + ' ' + statusDotId1}
        />
        <Checkbox
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
              ]}
              sublabel={<Sublabel id={sublabelId2} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId2} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId2 + ' ' + statusDotId2}
        />
        <RadioButton
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
              ]}
              sublabel={<Sublabel id={sublabelId3} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId3} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId3 + ' ' + statusDotId3}
        />
      </>
    );
  },
};

export const OnDark: Story = {
  render: args => {
    const sublabelId = 'sublabel-testid';
    const sublabelTexts: LabelText[] = [
      { text: 'normal sublabel with a pretty long text', type: 'normal' },
      { text: 'Semibold sublabel', type: 'semibold' },
    ];
    const statusDotId = 'statusdot-testid';

    return (
      <div style={{ padding: '3rem', backgroundColor: getColor('blueberry', 600) }}>
        <Input
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label', type: 'normal' },
                { text: 'semibold returns', type: 'semibold' },
                { text: 'another normal label', type: 'normal' },
              ]}
              sublabel={<Sublabel id={sublabelId} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          mode={'ondark'}
          aria-describedby={sublabelId + ' ' + statusDotId}
        />
      </div>
    );
  },
};

export const ChildrenAfterLabel: Story = {
  render: args => (
    <Input
      {...args}
      label={
        <Label
          labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]}
          afterLabelChildren={<Icon size={IconSize.XSmall} svgIcon={HelpSign} />}
        />
      }
    />
  ),
};

export const LabelAsString: Story = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  render: ({ showIcon, ...rest }) => (
    <>
      <Input {...rest} label={'Test label'} />
      <Textarea {...rest} label={'Test label'} />
      <Select {...rest} label={'Test label'}>
        <option value={'Option 1'}>{'Option 1'}</option>
        <option value={'Option 2'}>{'Option 2'}</option>
        <option value={'Option 3'}>{'Option 3'}</option>
      </Select>
      <Checkbox {...rest} label={'Test label'} />
      <RadioButton {...rest} label={'Test label'} />
    </>
  ),
};
