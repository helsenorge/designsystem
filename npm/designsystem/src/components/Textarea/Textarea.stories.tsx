import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Textarea from './Textarea';
import { FormMode } from '../../constants';
import Label from '../Label/Label';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Textarea',
  component: Textarea,
  parameters: {
    docs: {
      description: {
        component:
          'Som innbygger vil jeg kunne skrive inn en lengre tekst og ha plass til å se flere linjer av teksten jeg skriver inn slik at jeg kan beholde oversikten og redigere teksten uten for stor ulempe.',
      },
    },
  },
  args: {
    maxCharacters: 150,
    maxText: 'tegn',
    width: undefined,
    mode: FormMode.onwhite,
    testId: '123-test',
    transparent: false,
    disabled: false,
    minRows: 3,
    maxRows: 15,
    grow: true,
    errorText: '',
    autoFocus: false,
    readOnly: false,
    autoComplete: '',
    name: 'textarea',
    required: false,
  },
  argTypes: {
    placeholder: {
      control: 'text',
    },
    maxCharacters: {
      control: 'number',
    },
    maxText: {
      control: 'text',
    },
    width: {
      control: 'number',
    },
    mode: {
      control: 'select',
      options: FormMode,
    },
    defaultValue: {
      control: 'text',
    },
    testId: {
      control: 'text',
    },
    transparent: {
      control: 'boolean',
    },
    disabled: {
      control: 'boolean',
    },
    minRows: {
      control: 'number',
    },
    maxRows: {
      control: 'number',
    },
    grow: {
      control: 'boolean',
    },
    errorText: {
      control: 'text',
    },
    autoFocus: {
      control: 'boolean',
    },
    readOnly: {
      control: 'boolean',
    },
    autoComplete: {
      control: 'text',
    },
    name: {
      control: 'text',
    },
    required: {
      control: 'boolean',
    },
  },
} satisfies Meta<typeof Textarea>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => <Textarea {...args} label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />} />,
};

export const MaxCharacters: Story = {
  render: args => (
    <>
      <Textarea
        {...args}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        maxCharacters={50}
        marginBottom
        width={50}
      />
      <Textarea
        {...args}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        maxCharacters={100}
        width={100}
      />
      <Textarea
        {...args}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        maxCharacters={100}
        defaultValue="test"
        width={100}
      />
    </>
  ),
};
