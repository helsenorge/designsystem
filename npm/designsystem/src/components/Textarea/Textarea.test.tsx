import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Textarea from './Textarea';
import { FormMode } from '../../constants';
import Label from '../Label';

describe('Gitt at Textarea skal vises', (): void => {
  describe('Når Textarea rendres', (): void => {
    test('Så vises Textarea', (): void => {
      const { container } = render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} />);

      const label = screen.getByText('Skriv din historie');
      expect(label).toBeVisible();

      const input = screen.getByRole('textbox');
      expect(input).toBeVisible();
      expect(input.className).toBe('input-container__input');
    });
  });

  describe('Når disabled er true', (): void => {
    test('Så vises Textarea som disabled', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} disabled />);

      const input = screen.getByRole('textbox');
      expect(input).toBeDisabled();
    });
  });

  describe('Når FormMode er onblueberry', (): void => {
    test('Så vises Textarea med onblueberry styling', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} mode={FormMode.onblueberry} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper?.className).toBe('input-container input-container--on-blueberry');
    });
  });
  describe('Når mode er ondark', (): void => {
    test('Så vises Textarea med ondark styling', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} mode={FormMode.ondark} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper?.className).toBe('input-container input-container--on-dark');
    });
  });

  describe('Når textarea har en default value', (): void => {
    test('Så skal default verdien vises', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} defaultValue="Min historie" />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveValue('Min historie');
    });
  });

  describe('Når textarea har en placeholder value', (): void => {
    test('Så skal placeholder verdien vises', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie' }]} />} placeholder="Fin placeholder" />);
      const input = screen.getByPlaceholderText('Fin placeholder');
      expect(input).toBeVisible();
    });
  });

  describe('Når textarea skal vises med custom textareaId', (): void => {
    test('Så har textarea id-en', (): void => {
      render(<Textarea textareaId="testing" />);

      const textarea = screen.getByRole('textbox');
      expect(textarea).toHaveAttribute('id', 'testing');
    });
  });

  describe('Når autoFocus er satt', (): void => {
    test('Så er textarea focus', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} autoFocus />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toHaveFocus();
    });
  });

  describe('Når disabled er satt', (): void => {
    test('Så er textarea disabled', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} disabled />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toBeDisabled();
    });
  });

  describe('Når readOnly er satt', (): void => {
    test('Så er textarea readOnly', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} readOnly />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toHaveAttribute('readonly', '');
    });
  });

  describe('Når autoComplete er satt til on', (): void => {
    test('Så har textarea autoComplete=on', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} autoComplete="on" />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toHaveAttribute('autoComplete', 'on');
    });
  });

  describe('Når name-prop er satt', (): void => {
    test('Så har textarea riktig name', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} name="custom-name" />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toHaveAttribute('name', 'custom-name');
    });
  });

  describe('Når placeholder-prop er satt', (): void => {
    test('Så har textarea riktig placeholder', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} placeholder="custom-placeholder" />);

      const textarea = screen.getByPlaceholderText('custom-placeholder');
      expect(textarea).toBeVisible();
    });
  });

  describe('Når defaultValue-prop er satt', (): void => {
    test('Så har textarea riktig value', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} defaultValue="custom-value" />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toHaveValue('custom-value');
    });
  });

  describe('Når required er satt', (): void => {
    test('Så er textarea required', (): void => {
      render(<Textarea label={<Label labelTexts={[{ text: 'En fin label' }]} />} required />);

      const textarea = screen.getByLabelText('En fin label');
      expect(textarea).toBeRequired();
    });
  });

  describe('Når antall tegn skal vises', (): void => {
    describe('Når man skriver', (): void => {
      test('Så endres teksten og antall tegn', async (): Promise<void> => {
        render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie her' }]} />} maxCharacters={50} />);

        const input = screen.getByLabelText('Skriv din historie her');

        expect(screen.getByText('0/50 tegn')).toBeVisible();

        await userEvent.type(input, 'Jeg tester teksten her.');

        expect(input).toHaveValue('Jeg tester teksten her.');
        expect(screen.getByText('23/50 tegn')).toBeVisible();
      });
    });

    describe('Når man skriver for mange tegn', (): void => {
      test('Så indikeres det at input er ugyldig', async (): Promise<void> => {
        render(<Textarea label={<Label labelTexts={[{ text: 'Skriv din historie her' }]} />} maxCharacters={10} />);

        const input = screen.getByLabelText('Skriv din historie her');

        expect(screen.getByText('0/10 tegn')).toBeVisible();

        await userEvent.type(input, 'Jeg tester teksten her.');

        expect(input).toHaveValue('Jeg tester teksten her.');
        expect(screen.getByText('23/10 tegn')).toBeVisible();
        expect(input).toHaveAttribute('aria-invalid', 'true');
      });
    });
  });
});
