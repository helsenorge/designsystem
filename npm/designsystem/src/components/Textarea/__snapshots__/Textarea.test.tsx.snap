// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Textarea skal vises Når Textarea rendres Så vises Textarea 1`] = `
<div>
  <div
    class="error-wrapper"
  >
    <div
      class="textarea"
      data-analyticsid="textarea"
    >
      <div
        class="textarea__label-wrapper"
      >
        <label
          for="test"
        >
          Skriv din historie
        </label>
      </div>
      <div
        class="content-wrapper"
      >
        <textarea
          aria-invalid="false"
          class="content-wrapper__input"
          id="test"
          rows="3"
        />
      </div>
    </div>
  </div>
</div>
`;
